File: CherryPlayServer/Core/Entities/PlaybackState.cs
using CherryPlayServer.Core.Enums;

namespace CherryPlayServer.Core.Entities;

public class PlaybackState
{
    public string? CurrentTrackId { get; set; }
    public PlaybackStatus Status { get; set; } = PlaybackStatus.Idle;
    public double Position { get; set; }
    public double Duration { get; set; }
    public double Volume { get; set; } = 0.8;
    public PlaybackMode Mode { get; set; } = PlaybackMode.Preparation;
    public List<string> PlayedTrackIds { get; set; } = [];
    public List<string> DisabledTrackIds { get; set; } = [];
    public List<string> DisabledGroupIds { get; set; } = [];
    public DateTime? SessionStartedAt { get; set; }
    public DateTime LastUpdatedAt { get; set; } = DateTime.UtcNow;

    public PlaybackState Clone()
    {
        return new PlaybackState
        {
            CurrentTrackId = CurrentTrackId,
            Status = Status,
            Position = Position,
            Duration = Duration,
            Volume = Volume,
            Mode = Mode,
            PlayedTrackIds = PlayedTrackIds.ToList(),
            DisabledTrackIds = DisabledTrackIds.ToList(),
            DisabledGroupIds = DisabledGroupIds.ToList(),
            SessionStartedAt = SessionStartedAt,
            LastUpdatedAt = LastUpdatedAt
        };
    }
}

File: CherryPlayList/src/shared/contracts/playbackState.ts
import type { StorePlaybackStatus } from '../audio/playback/bindPlaybackEngineToStore';

import type { AimpPlaybackStatus } from './aimp';

/**
 * Playback status on the wire (REST / SignalR).
 * Matches CONTRACTS.md §6.3 and server `PlaybackStatus` enum.
 */
export type PlaybackWireStatus = 'idle' | 'playing' | 'paused' | 'ended';

export interface PlaybackStateDto {
  currentTrackId: string | null;
  status: PlaybackWireStatus;
  position: number;
  duration: number;
  volume: number;
  mode: 'preparation' | 'session';
  playedTrackIds: string[];
  disabledTrackIds: string[];
  disabledGroupIds: string[];
  lastUpdatedAt: string;
}

/**
 * Maps local player store statuses to the wire contract.
 * Transitional statuses (`loading`, `buffering`) are published as `playing`.
 */
export function mapStoreStatusToWireStatus(status: StorePlaybackStatus): PlaybackWireStatus {
  switch (status) {
    case 'playing':
    case 'loading':
    case 'buffering':
      return 'playing';
    case 'paused':
      return 'paused';
    case 'ended':
      return 'ended';
    case 'error':
    case 'idle':
    default:
      return 'idle';
  }
}

/** Maps AIMP playback snapshot status to the wire contract. */
export function mapAimpPlaybackStatusToWireStatus(
  status: AimpPlaybackStatus | null | undefined,
): PlaybackWireStatus {
  switch (status) {
    case 'playing':
      return 'playing';
    case 'paused':
      return 'paused';
    case 'stopped':
    default:
      return 'idle';
  }
}
